src/Chainweb/BlockHeaderDB.hs

```diff
@@ -324,7 +330,7 @@ pruneForks logger cdb callback limit = do
             -- Delete nodes which parent isn't in the filter or is in the
             -- falsePositiveSet
             parentIsMarked <- stToIO $ BF.elem p marked
-            if not (isGenesisBlockHeader h) && (not parentIsMarked || HS.member p falsePositiveSet)
+            if not (isGenesisBlockHeader h) && (not parentIsMarked || HS.member p falsePositiveSet || not isPayloadMarked)
```

Comment:
Laziness question: if the `HS.member` call is true, will the `IO` involved in `isPayloadMarked` be ran? I'm trying to figure out which is better to place first.

Reply:
Binding in `IO` is generally strict. Lazy `IO` is only used for "true" input/output functions. I would expected that `stToIO` evaluates its argument strictly.

Reply:
I guess it depends on whether `BF.elem` is strict or returns a thunk. I would expect the former, but not sure.

Reply:
In that case, it might be a tiny bit more performant to put the `isPayloadMarked` check before the `HS.member`.

Reply:
I tried to refactor but my attempts led to more complex control flow. All of these checks should be pretty cheap (no IO involved), so I don't think optimizing this is worth the effort.